source/cbor.c

```diff
@@ -0,0 +1,774 @@
+    } else {
+        PyErr_Format(PyExc_ValueError, "Not supported type %R", (PyObject *)Py_TYPE(py_object));
+    }
+
+    Py_RETURN_NONE;
+}
+
+PyObject *aws_py_cbor_encoder_write_py_list(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *py_object;
+    PyObject *py_capsule;
+    if (!PyArg_ParseTuple(args, "OO", &py_capsule, &py_object)) {
+        return NULL;
+    }
+    struct encoder_binding *binding = PyCapsule_GetPointer(py_capsule, s_capsule_name_cbor_encoder);
+    if (!binding) {
+        return NULL;
+    }
+    return s_cbor_encoder_write_pylist(binding, py_object);
+}
+
+PyObject *aws_py_cbor_encoder_write_py_dict(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *py_object;
+    PyObject *py_capsule;
+    if (!PyArg_ParseTuple(args, "OO", &py_capsule, &py_object)) {
+        return NULL;
+    }
+    struct encoder_binding *binding = PyCapsule_GetPointer(py_capsule, s_capsule_name_cbor_encoder);
+    if (!binding) {
+        return NULL;
+    }
+    return s_cbor_encoder_write_pydict(binding, py_object);
+}
+
+PyObject *aws_py_cbor_encoder_write_data_item(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *py_object;
+    PyObject *py_capsule;
+    if (!PyArg_ParseTuple(args, "OO", &py_capsule, &py_object)) {
+        return NULL;
+    }
+    struct encoder_binding *binding = PyCapsule_GetPointer(py_capsule, s_capsule_name_cbor_encoder);
+    if (!binding) {
+        return NULL;
+    }
+    return s_cbor_encoder_write_pyobject(binding, py_object);
+}
+
+/*******************************************************************************
+ * DECODE
+ ******************************************************************************/
+
+static const char *s_capsule_name_cbor_decoder = "aws_cbor_decoder";
+
+static struct aws_cbor_decoder *s_cbor_decoder_from_capsule(PyObject *py_capsule) {
+    return PyCapsule_GetPointer(py_capsule, s_capsule_name_cbor_decoder);
+}
+/* Runs when GC destroys the capsule */
+static void s_cbor_decoder_capsule_destructor(PyObject *py_capsule) {
+    struct aws_cbor_decoder *decoder = s_cbor_decoder_from_capsule(py_capsule);
+    aws_cbor_decoder_destroy(decoder);
+}
+
+PyObject *aws_py_cbor_decoder_new(PyObject *self, PyObject *args) {
+    (void)self;
+    /* The python object will keep the src alive from python. */
+    struct aws_byte_cursor src; /* s# */
+
+    if (!PyArg_ParseTuple(args, "s#", &src.ptr, &src.len)) {
+        return NULL;
+    }
+
+    struct aws_cbor_decoder *decoder = aws_cbor_decoder_new(aws_py_get_allocator(), src);
+    AWS_ASSERT(decoder != NULL);
+    PyObject *py_capsule = PyCapsule_New(decoder, s_capsule_name_cbor_decoder, s_cbor_decoder_capsule_destructor);
+    if (!py_capsule) {
+        aws_cbor_decoder_destroy(decoder);
+        return NULL;
+    }
+
+    return py_capsule;
+}
+
+#define S_GET_DECODER()                                                                                                \
```

Comment:
nit: but im not a big fan of macros inserting random code into functions, can we make macros more like a generic function that returns a value rathere then relying on side effects